### PR TITLE
protocols: add hyprland_focus_grab_v1 implementation

### PR DESCRIPTION
**Pull request by outfoxxed**
#### Describe your PR, what does it fix/add?
Implements hyprland_focus_grab_v1 as described in https://github.com/hyprwm/hyprland-protocols/pull/7.

This implementation has been tested somewhat thoroughly against my client implementation.
(All the usual cases, correct usage, destroying surfaces semi randomly, overlapping grabs, client dies, etc)

The clientside implementation can be found here: https://git.outfoxxed.me/outfoxxed/quickshell/src/branch/hyprland-focus-grab

Very simple tester:
```qml
import Quickshell
import Quickshell.Hyprland
import QtQuick.Controls

ShellRoot {
  FloatingWindow {
    id: window
    Button {
      anchors.centerIn: parent
      text: grab.active ? "Remove exclusive focus" : "Take exclusive focus"
      onClicked: grab.active = !grab.active
    }

    HyprlandFocusGrab {
      id: grab
      windows: [ window ]
    }
  }
}
```

This can be run with `quickshell -p tester.qml`

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
I can't test touch so I'm not sure if that works correctly.

#### Is it ready for merging, or does it need work?
Yes, also after merging https://github.com/hyprwm/hyprland-protocols/pull/7 the nix lockfile will need to be updated for it to build.

### CODE REVIEW

File: src/protocols/FocusGrab.cpp [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+            break;
+        }
+    }
+
+    if (surface) {
+        const auto KEYBOARD     = wlr_seat_get_keyboard(g_pCompositor->m_sSeat.seat);
+        uint32_t   keycodes[32] = {0};
+
+        wlr_seat_keyboard_enter(g_pCompositor->m_sSeat.seat, m_mSurfaces.begin()->first, keycodes, 0, &KEYBOARD->modifiers);
+    } else {
```
Comment (vaxerski) [inline src/protocols/FocusGrab.cpp:269]: no {} around short ifs

File: src/protocols/FocusGrab.cpp [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+        // Ensure surfaces under the mouse when the grab ends get focus.
+        if (hadGrab)
+            g_pInputManager->refocus();
+    }
+}
+
+void CFocusGrab::addSurface(wlr_surface* surface) {
+    auto iter = m_mSurfaces.find(surface);
+    if (iter == m_mSurfaces.end()) {
```
Comment (vaxerski) [inline src/protocols/FocusGrab.cpp:231]: .

File: src/protocols/FocusGrab.cpp [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        m_bGrabActive = false;
+        hyprListener_pointerGrabStarted.removeCallback();
+        hyprListener_keyboardGrabStarted.removeCallback();
+        hyprListener_touchGrabStarted.removeCallback();
+
+        // Only clear grabs that belong to this focus grab. When superseded by another grab
+        // or xdg_popup grab we might not own the current grab.
+
+        bool hadGrab = false;
+        if (g_pCompositor->m_sSeat.seat->pointer_state.grab == &this->m_sPointerGrab) {
```
Comment (vaxerski) [inline src/protocols/FocusGrab.cpp:203]: no this->

File: src/protocols/FocusGrab.cpp [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+            surface = surf;
+            break;
+        }
+    }
+
+    if (surface) {
+        const auto KEYBOARD     = wlr_seat_get_keyboard(g_pCompositor->m_sSeat.seat);
+        uint32_t   keycodes[32] = {0};
+
+        wlr_seat_keyboard_enter(g_pCompositor->m_sSeat.seat, m_mSurfaces.begin()->first, keycodes, 0, &KEYBOARD->modifiers);
```
Comment (vaxerski) [inline src/protocols/FocusGrab.cpp:268]: g_pCompositor->focusSurface

File: src/protocols/FocusGrab.cpp [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+    .motion      = focus_grab_pointer_motion,
+    .button      = focus_grab_pointer_button,
+    .axis        = focus_grab_pointer_axis,
+    .frame       = focus_grab_pointer_frame,
+    .cancel      = focus_grab_pointer_cancel,
+};
+
+static void focus_grab_keyboard_enter(wlr_seat_keyboard_grab* grab, wlr_surface* surface, const uint32_t keycodes[], size_t num_keycodes, const wlr_keyboard_modifiers* modifiers) {
+    if (static_cast<CFocusGrab*>(grab->data)->isSurfaceComitted(surface))
+        wlr_seat_keyboard_enter(grab->seat, surface, keycodes, num_keycodes, modifiers);
```
Comment (vaxerski) [inline src/protocols/FocusGrab.cpp:60]: g_pCompositor->focusSurface

File: src/protocols/FocusGrab.cpp [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+#include "Compositor.hpp"
+#include <hyprland-focus-grab-v1.hpp>
+#include <managers/input/InputManager.hpp>
+#include <cstdint>
+#include <memory>
+#include <wayland-server.h>
+
+static void focus_grab_pointer_enter(wlr_seat_pointer_grab* grab, wlr_surface* surface, double sx, double sy) {
+    if (static_cast<CFocusGrab*>(grab->data)->isSurfaceComitted(surface))
+        wlr_seat_pointer_enter(grab->seat, surface, sx, sy);
```
Comment (vaxerski) [inline src/protocols/FocusGrab.cpp:11]: g_pCompositor->focusSurface

File: src/protocols/FocusGrab.cpp [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+    else
+        wlr_seat_pointer_clear_focus(grab->seat);
+}
+
+static void focus_grab_pointer_clear_focus(wlr_seat_pointer_grab* grab) {
+    wlr_seat_pointer_clear_focus(grab->seat);
```
Comment (vaxerski) [inline src/protocols/FocusGrab.cpp:17]: g_pCompositor->focusSurface

File: src/desktop/Popup.cpp [hunk 8]
```diff
@@ -36,6 +36,10 @@ CPopup::~CPopup() {
     hyprListener_destroyPopup.removeCallback();
 }
 
+wlr_xdg_popup* CPopup::wlr() {
+    return this->m_pWLR;
```
[the remark this answers is not in the document]
Comment (outfoxxed) [reply src/desktop/Popup.cpp]: this bit got force pushed out when I removed the overcomplicated kb surface entry

File: src/protocols/FocusGrab.cpp (hunk 5 above)
Comment (outfoxxed) [reply src/protocols/FocusGrab.cpp:60]: recursive

**Review by vaxerski — APPROVED** [review]
lgtm

**Pull request merged**